src/main/java/org/wildfly/security/credential/store/CredentialStoreSpi.java

```diff
@@ -125,4 +143,78 @@ public boolean isInitialized() {
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * Resolves master password based on {@code commandSpec} using {@code attributes}
+     * If this method is not able to detect provider based on {@code commandSpec} it simply wraps {@code commandSpec} as
+     * password and generates {@link PasswordCredential} based on {@link ClearPasswordSpec}.
+     * @param commandSpec password command specification
+     * @param credentialType type of credential returned
+     * @param attributes to use while resolving credential (can be {@code null})
+     * @param <C> the type of credential which will be returned
+     * @return resolved credential of {@code credentialType}
+     * @throws CredentialStoreException if anything goes wrong
+     */
+    protected <C extends Credential> C resolveMasterCredential(String commandSpec, Class<C> credentialType, Map<String, String> attributes)
+            throws CredentialStoreException {
+        Assert.assertNotNull(commandSpec);
+        Assert.assertNotNull(credentialType);
+        final SecurityManager sm = System.getSecurityManager();
+        if (sm != null) {
+            sm.checkPermission(CredentialStorePermission.LOAD_EXTERNAL_STORE_PASSWORD);
+        }
+        if (commandSpec.trim().startsWith("{" + ExecPasswordStore.SUPPORTED_CMD_TYPE)) {
```

Comment:
I assume not called often so not a major issue but maybe the result of the first call to trim should be cached - at the moment it is repeated for each if.